File: src/geoseal.py
"""
GeoSeal: Geometric Access Control Kernel (Python Reference Implementation)
==========================================================================

Immune-like dynamics for Spiralverse RAG latent space using hyperbolic
geometry and phase-discipline to quarantine adversarial or off-grammar
retrievals.

Core mechanisms:
- Phase validity -> repulsion amplification (null phase = 2.0x, wrong = 1.5x + deviation)
- Per-neighbor suspicion counters (temporal integration, filters transient flukes)
- Spatial consensus threshold (3+ neighbors agreeing = quarantine mode)
- Second-stage amplification (force x 1.5 when quarantined)
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

# ---------------------------------------------------------------------------
# Sacred Tongues phase mapping
# ---------------------------------------------------------------------------

TONGUE_PHASES: Dict[str, float] = {
    "KO": 0.0,
    "AV": math.pi / 3,
    "RU": 2 * math.pi / 3,
    "CA": math.pi,
    "UM": 4 * math.pi / 3,
    "DR": 5 * math.pi / 3,
}

# Suspicion / quarantine parameters (defaults; tune per deployment via GeoSealConfig)
SUSPICION_DECAY = 0.5
SUSPICION_THRESHOLD = 3
QUARANTINE_CONSENSUS = 3
TRUST_DENOMINATOR = 20.0


@dataclass(frozen=True)
class GeoSealConfig:
    """Tunable GeoSeal parameters. Defaults preserve historical behavior exactly.

    Lets deployments (and hyperparameter sweeps) tune suspicion dynamics
    without editing module constants.
    """

    suspicion_decay: float = SUSPICION_DECAY
    suspicion_threshold: float = SUSPICION_THRESHOLD
    quarantine_consensus: int = QUARANTINE_CONSENSUS
    trust_denominator: float = TRUST_DENOMINATOR
    ball_radius: float = 0.99

    def __post_init__(self) -> None:
        if self.suspicion_decay < 0:
            raise ValueError(f"suspicion_decay must be >= 0, got {self.suspicion_decay}")
        if self.suspicion_threshold <= 0:
            raise ValueError(f"suspicion_threshold must be > 0, got {self.suspicion_threshold}")
        if self.quarantine_consensus < 1:
            raise ValueError(f"quarantine_consensus must be >= 1, got {self.quarantine_consensus}")
        if self.trust_denominator <= 0:
            raise ValueError(f"trust_denominator must be > 0, got {self.trust_denominator}")
        if not 0.0 < self.ball_radius < 1.0:
            raise ValueError(f"ball_radius must be in (0, 1), got {self.ball_radius}")


DEFAULT_CONFIG = GeoSealConfig()


# ---------------------------------------------------------------------------
# Agent
# ---------------------------------------------------------------------------


@dataclass
class Agent:
    """An agent in the GeoSeal swarm (retrieval chunk, tongue, or memory)."""

    id: str
    position: List[float]
    phase: Optional[float]
    tongue: Optional[str] = None
    suspicion_count: Dict[str, float] = field(default_factory=dict)
    is_quarantined: bool = False
    trust_score: float = 1.0


# ---------------------------------------------------------------------------
# Poincaré ball primitives
# ---------------------------------------------------------------------------


def _norm_sq(v: List[float]) -> float:
    return sum(x * x for x in v)


def _norm(v: List[float]) -> float:
    return math.sqrt(_norm_sq(v))


def hyperbolic_distance(u: List[float], v: List[float]) -> float:
    """Compute hyperbolic distance in Poincaré ball model.

    d_H(u, v) = arcosh(1 + 2 * ||u - v||² / ((1 - ||u||²)(1 - ||v||²)))

    Raises ValueError on dimension mismatch (zip would silently truncate;
    the TypeScript port throws RangeError — behavior is now consistent).
    """
    if len(u) != len(v):
        raise ValueError(f"dimension mismatch: len(u)={len(u)} != len(v)={len(v)}")
    diff = [a - b for a, b in zip(u, v)]
    diff_norm_sq = _norm_sq(diff)
    u_norm_sq = _norm_sq(u)
    v_norm_sq = _norm_sq(v)

    denom = (1 - u_norm_sq) * (1 - v_norm_sq)
    if denom <= 0:
        return float("inf")

    arg = 1 + 2 * diff_norm_sq / denom
    return math.acosh(max(1.0, arg))


def phase_deviation(phase1: Optional[float], phase2: Optional[float]) -> float:
    """Compute normalized phase deviation in [0, 1].

    None phase = maximum deviation (1.0).
    """
    if phase1 is None or phase2 is None:
        return 1.0

    diff = abs(phase1 - phase2)
    if diff > math.pi:
        diff = 2 * math.pi - diff

    return diff / math.pi


def clamp_to_ball(v: List[float], r_max: float = 0.99) -> List[float]:
    """Clamp a vector to the Poincaré ball of radius r_max."""
    n = _norm(v)
    if n >= r_max:
        scale = r_max / n
        return [x * scale for x in v]
    return list(v)


# ---------------------------------------------------------------------------
# Repulsion force
# ---------------------------------------------------------------------------


def compute_repel_force(
    agent_a: Agent,
    agent_b: Agent,
    base_strength: float = 1.0,
) -> Tuple[List[float], float, bool]:
    """Core GeoSeal repulsion computation.

    Returns (force_vector, amplification, anomaly_flag).
    """
    d_h = hyperbolic_distance(agent_a.position, agent_b.position)
    base_repulsion = base_strength / (d_h + 1e-6)

    amplification = 1.0
    anomaly_flag = False

    if agent_b.phase is None:
        amplification = 2.0
        anomaly_flag = True
    elif agent_a.phase is not None:
        deviation = phase_deviation(agent_a.phase, agent_b.phase)
        if d_h < 1.0 and deviation > 0.5:
            amplification = 1.5 + deviation
            anomaly_flag = True

    if agent_b.is_quarantined:
        amplification *= 1.5

    direction = [a - b for a, b in zip(agent_a.position, agent_b.position)]
    force = [d * base_repulsion * amplification for d in direction]

    return force, amplification, anomaly_flag


# ---------------------------------------------------------------------------
# Suspicion tracking
# ---------------------------------------------------------------------------


def update_suspicion(
    agent: Agent,
    neighbor_id: str,
    is_anomaly: bool,
    config: Optional[GeoSealConfig] = None,
) -> None:
    """Update suspicion counters and quarantine status."""
    cfg = config or DEFAULT_CONFIG
    if is_anomaly:
        agent.suspicion_count[neighbor_id] = agent.suspicion_count.get(neighbor_id, 0) + 1
    else:
        agent.suspicion_count[neighbor_id] = max(0, agent.suspicion_count.get(neighbor_id, 0) - cfg.suspicion_decay)

    suspicious_neighbors = sum(1 for c in agent.suspicion_count.values() if c >= cfg.suspicion_threshold)
    agent.is_quarantined = suspicious_neighbors >= cfg.quarantine_consensus

    total_suspicion = sum(agent.suspicion_count.values())
    agent.trust_score = max(0, 1.0 - total_suspicion / cfg.trust_denominator)


# ---------------------------------------------------------------------------
# Swarm dynamics
# ---------------------------------------------------------------------------


def swarm_step(
    agents: List[Agent],
    drift_rate: float = 0.01,
    config: Optional[GeoSealConfig] = None,
) -> List[Agent]:
    """Run one swarm update step for all agents.

    Raises ValueError if agents carry positions of different dimensions —
    catching embedding mismatches here instead of deep in the force loop.
    """
    cfg = config or DEFAULT_CONFIG
    n = len(agents)
    if n == 0:
        return agents

    dim = len(agents[0].position)
    for a in agents:
        if len(a.position) != dim:
            raise ValueError(f"agent {a.id!r} has dimension {len(a.position)}, expected {dim} (from {agents[0].id!r})")

    for i in range(n):
        net_force = [0.0] * dim

        for j in range(n):
            if i == j:
                continue

            force, _amp, anomaly = compute_repel_force(agents[i], agents[j])
            for k in range(dim):
                net_force[k] += force[k]

            # Update suspicion on the TARGET (j) from the SOURCE (i)
            # When i flags j as anomalous, j's suspicion record grows
            update_suspicion(agents[j], agents[i].id, anomaly, cfg)

        # Apply force
        for k in range(dim):
            agents[i].position[k] += net_force[k] * drift_rate

        # Clamp to Poincaré ball
        agents[i].position = clamp_to_ball(agents[i].position, cfg.ball_radius)

    return agents


def run_swarm(
    agents: List[Agent],
    num_steps: int = 10,
    drift_rate: float = 0.01,
    config: Optional[GeoSealConfig] = None,
) -> List[Agent]:
    """Run multiple swarm steps."""
    for _ in range(num_steps):
        swarm_step(agents, drift_rate, config)
    return agents


# ---------------------------------------------------------------------------
# Metrics
# ---------------------------------------------------------------------------


@dataclass
class GeoSealMetrics:
    """Benchmarking metrics for a GeoSeal run."""

    time_to_isolation: int
    boundary_norm: float
    suspicion_consensus: float
    collateral_flags: int
    final_trust_scores: Dict[str, float]


def compute_metrics(
    agents: List[Agent],
    rogue_id: str,
    config: Optional[GeoSealConfig] = None,
) -> GeoSealMetrics:
    """Compute GeoSeal metrics for a completed swarm run."""
    cfg = config or DEFAULT_CONFIG
    rogue = next((a for a in agents if a.id == rogue_id), None)
    if rogue is None:
        raise ValueError(f"Rogue agent not found: {rogue_id}")

    norm = _norm(rogue.position)

    suspicious = sum(1 for c in rogue.suspicion_count.values() if c >= cfg.suspicion_threshold)
    total_neighbors = len(rogue.suspicion_count)
    consensus = suspicious / total_neighbors if total_neighbors > 0 else 0.0

    collateral = sum(1 for a in agents if a.is_quarantined and a.phase is not None)

    trust_scores = {a.id: a.trust_score for a in agents}

    return GeoSealMetrics(
        time_to_isolation=len(agents) if rogue.is_quarantined else -1,
        boundary_norm=norm,
        suspicion_consensus=consensus,
        collateral_flags=collateral,
        final_trust_scores=trust_scores,
    )

File: src/geoseal.ts
/**
 * @file geoseal.ts
 * @module geoseal
 * @layer Layer 9, Layer 12, Layer 13
 * @version 1.0.0
 *
 * GeoSeal: Geometric Access Control Kernel
 *
 * Turns vector RAG from a passive similarity-matcher into an active immune
 * system that detects, quarantines, and reweights adversarial or off-grammar
 * retrievals using hyperbolic geometry and phase-discipline consensus.
 *
 * Core mechanisms:
 * - Phase validity -> repulsion amplification (null phase = 2.0x, wrong = 1.5x + deviation)
 * - Per-neighbor suspicion counters (temporal integration, filters transient flukes)
 * - Spatial consensus threshold (3+ neighbors agreeing = quarantine mode)
 * - Second-stage amplification (force x 1.5 when quarantined)
 *
 * Integration: Uses Poincaré ball primitives from harmonic/hyperbolic.ts.
 */

import {
  hyperbolicDistance,
  phaseDeviation,
  clampToBall,
} from './harmonic/hyperbolic.js';

// ═══════════════════════════════════════════════════════════════
// Sacred Tongues phase mapping
// ═══════════════════════════════════════════════════════════════

/** Fixed phases for the 6 Sacred Tongues (evenly spaced on unit circle) */
export const TONGUE_PHASES: Record<string, number> = {
  KO: 0.0, // Kor'aelin - Control/orchestration
  AV: Math.PI / 3, // Avali - Initialization/transport
  RU: (2 * Math.PI) / 3, // Runethic - Policy/authorization
  CA: Math.PI, // Cassisivadan - Encryption/compute
  UM: (4 * Math.PI) / 3, // Umbroth - Redaction/privacy
  DR: (5 * Math.PI) / 3, // Draumric - Authentication/integrity
};

// ═══════════════════════════════════════════════════════════════
// Agent type
// ═══════════════════════════════════════════════════════════════

export interface Agent {
  id: string;
  position: number[]; // Embedding vector (Poincaré ball)
  phase: number | null; // Tongue phase, or null if rogue
  tongue?: string; // Which Sacred Tongue
  suspicion_count: Map<string, number>; // Per-neighbor suspicion
  is_quarantined: boolean;
  trust_score: number; // 0.0 = untrusted, 1.0 = fully trusted
}

/** Create a new agent with sensible defaults */
export function createAgent(
  id: string,
  position: number[],
  phase: number | null,
  tongue?: string,
  trust_score: number = 0.5
): Agent {
  return {
    id,
    position: [...position],
    phase,
    tongue,
    suspicion_count: new Map(),
    is_quarantined: false,
    trust_score,
  };
}

// ═══════════════════════════════════════════════════════════════
// Repulsion force computation
// ═══════════════════════════════════════════════════════════════

export interface RepulsionResult {
  force: number[];
  amplification: number;
  anomaly_flag: boolean;
}

/**
 * Core GeoSeal repulsion force computation.
 *
 * Implements immune-like response: phase-mismatched agents near legitimate
 * agents experience amplified repulsive forces that push them toward the
 * manifold boundary.
 *
 * Amplification rules:
 * - Null phase (unknown/rogue): 2.0x
 * - Phase mismatch at close range (d_H < 1.0, deviation > 0.5): 1.5x + deviation
 * - Quarantined agents: additional 1.5x multiplier
 */
export function computeRepelForce(
  agent_a: Agent,
  agent_b: Agent,
  base_strength: number = 1.0
): RepulsionResult {
  const d_H = hyperbolicDistance(agent_a.position, agent_b.position);

  // Base repulsion: inversely proportional to distance
  const base_repulsion = base_strength / (d_H + 1e-6);

  // Compute phase-based amplification
  let amplification = 1.0;
  let anomaly_flag = false;

  if (agent_b.phase === null) {
    // Null phase (unknown/rogue) -> maximum amplification
    amplification = 2.0;
    anomaly_flag = true;
  } else if (agent_a.phase !== null) {
    // Both have phases, check for mismatch
    const deviation = phaseDeviation(agent_a.phase, agent_b.phase);

    // Close distance + high phase deviation = suspicious
    if (d_H < 1.0 && deviation > 0.5) {
      amplification = 1.5 + deviation;
      anomaly_flag = true;
    }
  }

  // Quarantined agents get further amplification
  if (agent_b.is_quarantined) {
    amplification *= 1.5;
  }

  // Force vector: direction away from agent_b
  const dim = agent_a.position.length;
  const force = new Array<number>(dim);
  for (let i = 0; i < dim; i++) {
    const direction = agent_a.position[i] - agent_b.position[i];
    force[i] = direction * base_repulsion * amplification;
  }

  return { force, amplification, anomaly_flag };
}

// ═══════════════════════════════════════════════════════════════
// Suspicion tracking and quarantine
// ═══════════════════════════════════════════════════════════════

/** Suspicion decay rate when no anomaly is detected */
export const SUSPICION_DECAY = 0.5;

/** Number of suspicion hits before a neighbor is considered suspicious */
export const SUSPICION_THRESHOLD = 3;

/** Number of suspicious neighbors required for quarantine */
export const QUARANTINE_CONSENSUS = 3;

/** Trust score denominator (total suspicion / this = trust loss) */
export const TRUST_DENOMINATOR = 20.0;

/**
 * Update suspicion counters and quarantine status for an agent.
 *
 * Requires 3+ neighbors with suspicion >= 3 for quarantine.
 * Suspicion decays by 0.5 per non-anomalous interaction (filters transient flukes).
 */
export function updateSuspicion(
  agent: Agent,
  neighbor_id: string,
  is_anomaly: boolean
): void {
  if (is_anomaly) {
    const current = agent.suspicion_count.get(neighbor_id) || 0;
    agent.suspicion_count.set(neighbor_id, current + 1);
  } else {
    // Decay suspicion if no anomaly detected
    const current = agent.suspicion_count.get(neighbor_id) || 0;
    agent.suspicion_count.set(neighbor_id, Math.max(0, current - SUSPICION_DECAY));
  }

  // Count how many neighbors are suspicious
  let suspicious_neighbors = 0;
  for (const count of agent.suspicion_count.values()) {
    if (count >= SUSPICION_THRESHOLD) suspicious_neighbors++;
  }

  // Quarantine threshold: QUARANTINE_CONSENSUS+ neighbors with high suspicion
  agent.is_quarantined = suspicious_neighbors >= QUARANTINE_CONSENSUS;

  // Update trust score (inverse of suspicion)
  let total_suspicion = 0;
  for (const count of agent.suspicion_count.values()) {
    total_suspicion += count;
  }
  agent.trust_score = Math.max(0, 1.0 - total_suspicion / TRUST_DENOMINATOR);
}

// ═══════════════════════════════════════════════════════════════
// Swarm dynamics
// ═══════════════════════════════════════════════════════════════

/**
 * Run one swarm update step for all agents.
 *
 * Computes pairwise repulsion forces, updates suspicion counters,
 * applies forces with drift rate, and clamps positions to Poincaré ball.
 *
 * @param agents - All agents in the swarm
 * @param drift_rate - Force application rate (default 0.01)
 * @returns Updated agents (new array, agents are mutated in place)
 */
export function swarmStep(agents: Agent[], drift_rate: number = 0.01): Agent[] {
  const n = agents.length;
  const dim = agents[0]?.position.length ?? 0;

  // Compute all pairwise forces
  for (let i = 0; i < n; i++) {
    const net_force = new Array<number>(dim).fill(0);

    for (let j = 0; j < n; j++) {
      if (i === j) continue;

      const result = computeRepelForce(agents[i], agents[j]);

      // Accumulate force
      for (let k = 0; k < dim; k++) {
        net_force[k] += result.force[k];
      }

      // Update suspicion on the TARGET (j) from the SOURCE (i)
      // When i flags j as anomalous, j's suspicion record grows
      updateSuspicion(agents[j], agents[i].id, result.anomaly_flag);
    }

    // Apply force with drift
    for (let k = 0; k < dim; k++) {
      agents[i].position[k] += net_force[k] * drift_rate;
    }

    // Clamp to Poincaré ball
    agents[i].position = clampToBall(agents[i].position, 0.99);
  }

  return agents;
}

/**
 * Run multiple swarm steps (convenience wrapper).
 *
 * @param agents - Initial agent set
 * @param num_steps - Number of swarm iterations
 * @param drift_rate - Force application rate
 * @returns Agents after all steps
 */
export function runSwarm(
  agents: Agent[],
  num_steps: number = 10,
  drift_rate: number = 0.01
): Agent[] {
  for (let step = 0; step < num_steps; step++) {
    swarmStep(agents, drift_rate);
  }
  return agents;
}
